src/Microsoft.AspNetCore.Antiforgery/Internal/AntiforgeryAuthenticationHandler.cs

```diff
@@ -0,0 +1,124 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Antiforgery.Internal;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features.Authentication;
+using Microsoft.AspNetCore.Http.Features.Authentication.Internal;
+
+namespace Microsoft.AspNetCore.Antiforgery.Internal
+{
+    public class AntiforgeryAuthenticationHandler : IAuthenticationHandler
+    {
+        public AntiforgeryAuthenticationHandler(IAntiforgery antiforgery)
+        {
+            Antiforgery = antiforgery;
+        }
+
+        protected IAntiforgery Antiforgery { get; }
+
+        public HttpContext HttpContext { get; private set; }
+
+        public IAuthenticationHandler PriorHandler { get; private set; }
+
+        public async Task InitializeAsync(HttpContext context)
+        {
+            HttpContext = context;
+
+            var authentication = GetAuthenticationFeature(HttpContext);
+
+            PriorHandler = authentication.Handler;
+            authentication.Handler = this;
+
+            if (authentication.User != null)
+            {
+                if (!await Antiforgery.IsRequestValidAsync(HttpContext))
+                {
+                    // Wipe out any existing principal if we can't validate this request.
+                    authentication.User = null;
+                    return;
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public async Task AuthenticateAsync(AuthenticateContext context)
+        {
+            if (PriorHandler != null)
+            {
+                await PriorHandler.AuthenticateAsync(context);
```

Comment:
Nevermind. If the middleware is placed out of order this may never even be called because the upstream provider would short circuit:
https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNetCore.Authentication/AuthenticationHandler.cs#L200-L233